upf/upf_ipfix.c

```diff
@@ -0,0 +1,958 @@
+  if (b)
+    upf_ipfix_export_send (vm, b, which, now);
+}
+
+void
+upf_ipfix_flush_callback_ip4 (u32 now)
+{
+  flush_record (FLOW_VARIANT_IP4, now);
+}
+
+void
+upf_ipfix_flush_callback_ip6 (u32 now)
+{
+  flush_record (FLOW_VARIANT_IP6, now);
+}
+
+/* void */
+/* upf_ipfix_flush_callback_l2 (void) */
+/* { */
+/*   flush_record (FLOW_VARIANT_L2); */
+/*   flush_record (FLOW_VARIANT_L2_IP4); */
+/*   flush_record (FLOW_VARIANT_L2_IP6); */
+/* } */
+
+/**
+ * @brief Flush accumulated data
+ * @param frm flow_report_main_t *
+ * @param fr flow_report_t *
+ * @param f vlib_frame_t *
+ *
+ * <em>Notes:</em>
+ * This function must simply return the incoming frame, or no template packets
+ * will be sent.
+ */
+vlib_frame_t *
+upf_ipfix_data_callback_ip4 (flow_report_main_t *frm, ipfix_exporter_t *exp,
+			     flow_report_t *fr, vlib_frame_t *f, u32 *to_next,
+			     u32 node_index)
+{
+  vlib_main_t *vm = vlib_get_main ();
+  u32 now = (u32) vlib_time_now (vm);
+  upf_ipfix_flush_callback_ip4 (now);
+  return f;
+}
+
+vlib_frame_t *
+upf_ipfix_data_callback_ip6 (flow_report_main_t *frm, ipfix_exporter_t *exp,
+			     flow_report_t *fr, vlib_frame_t *f, u32 *to_next,
+			     u32 node_index)
+{
+  vlib_main_t *vm = vlib_get_main ();
+  u32 now = (u32) vlib_time_now (vm);
+  upf_ipfix_flush_callback_ip6 (now);
+  return f;
+}
+
+/* vlib_frame_t * */
+/* upf_ipfix_data_callback_l2 (flow_report_main_t *frm, ipfix_exporter_t *exp, */
+/* 			    flow_report_t *fr, vlib_frame_t *f, u32 *to_next, */
+/* 			    u32 node_index) */
+/* { */
+/*   upf_ipfix_flush_callback_l2 (); */
+/*   return f; */
+/* } */
+
+static int
+upf_ipfix_template_add_del (u32 domain_id, u16 src_port,
+			    upf_ipfix_record_t flags,
+			    vnet_flow_data_callback_t * flow_data_callback,
+			    vnet_flow_rewrite_callback_t * rewrite_callback,
+			    bool is_add, u16 * template_id)
+{
+  ipfix_exporter_t *exp = &flow_report_main.exporters[0];
+  vnet_flow_report_add_del_args_t a = {
+    .rewrite_callback = rewrite_callback,
+    .flow_data_callback = flow_data_callback,
+    .is_add = is_add,
+    .domain_id = domain_id,
+    .src_port = src_port,
+    .opaque.as_uword = flags,
+  };
+  return vnet_flow_report_add_del (exp, &a, template_id);
+}
+
+static void upf_ipfix_export_entry (vlib_main_t * vm, flow_entry_t * f, flow_direction_t direction, u32 now);
+
+/*
+ * NTP rfc868 : 2 208 988 800 corresponds to 00:00  1 Jan 1970 GMT
+ */
+#define NTP_TIMESTAMP 2208988800LU
+
+static inline u32
+upf_ipfix_common_add (vlib_buffer_t * to_b, flow_entry_t * f, flow_direction_t direction,
+		      u16 offset)
+{
+  u16 start = offset;
+  upf_main_t *gtm = &upf_main;
+  upf_session_t *sx = pool_elt_at_index (gtm->sessions, f->session_index);
+
+  /* mobileIMSI */
+  /* the value can't be 255 or more due to limitations of pfcp.c, */
+  /* thus we use simpler encoding */
+  to_b->data[offset] = sx->imsi_len;
+  clib_memcpy_fast (to_b->data + offset + 1, sx->imsi, sx->imsi_len);
```

Comment:
I gave this a whirl out of curiosity for the functionality and while pretty much everything works out of the box the imsi ipfix field is not decoded by wireshark or any ipfix collector I've tried. As far as I could see the mobileImsi is sent in the [ipfix payload as a string](https://www.iana.org/assignments/ipfix/ipfix.xhtml) so, considering the imsi has a length of 15 codepoints the provided buffer should be 16 (imsi + null terminator).  `sx->imsi` is currently stored as 8 bytes though.
I guess here you need to format the mobile identify and allocate a 16 buffer for this to work:

```c
diff --git a/upf/upf_ipfix.c b/upf/upf_ipfix.c
index e493271..4756036 100644
--- a/upf/upf_ipfix.c
+++ b/upf/upf_ipfix.c
@@ -470,9 +470,15 @@ upf_ipfix_common_add (vlib_buffer_t * to_b, flow_entry_t * f, flow_direction_t d
   /* mobileIMSI */
   /* the value can't be 255 or more due to limitations of pfcp.c, */
   /* thus we use simpler encoding */
-  to_b->data[offset] = sx->imsi_len;
-  clib_memcpy_fast (to_b->data + offset + 1, sx->imsi, sx->imsi_len);
-  offset += sx->imsi_len + 1;
+
+  // note: imsi string is 15 code points and is currently stored as an 8 bit array (u8* -> sizeof(sx->imsi)=8).
+  // the ipfix rfc state that the imsi is encoded as a string (unicode uses 1 byte per codepoint for the first 128 codepoints). Hence when transfering the imsi in the ipfix payload
+  // we need to pre-allocate a buffer of size 16 before calling format_mobile_identity (15 codepoints + null terminator). This is equivalent to  2 * sx->imsi_len + 1.
+
+  to_b->data[offset] = sx->imsi_len * 2;
+  u8 *imsi = format (NULL, "%U", format_mobile_identity, sx->imsi, sx->imsi_len);
+  clib_memcpy_fast (to_b->data + offset + 1, imsi, 2 * sx->imsi_len);
+  offset += 2 * sx->imsi_len + 1;
```

Snippet quality might not be the best one ever but works as a PoC for me :)

Reply:
Btw I also got a segmentation fault on the vpp startup, gdb pointed out at the inability of allocating memory for the flow table. I had to reduce its size for this to work:

```
diff --git a/upf/flowtable.h b/upf/flowtable.h
index e535d87..467242d 100644
--- a/upf/flowtable.h
+++ b/upf/flowtable.h
@@ -207,7 +207,7 @@ typedef struct
  * hashtable is configured to alloc (NUM_BUCKETS * CLIB_CACHE_LINE_BYTES) Bytes
  * with (flow_count / (BIHASH_KVP_PER_PAGE / 2)) Buckets
  */
-#define FM_POOL_COUNT_LOG2 22
+#define FM_POOL_COUNT_LOG2 15
 #define FM_POOL_COUNT (1 << FM_POOL_COUNT_LOG2)
 #define FM_NUM_BUCKETS (1 << (FM_POOL_COUNT_LOG2 - (BIHASH_KVP_PER_PAGE / 2)))
 #define FM_MEMORY_SIZE (FM_NUM_BUCKETS * CLIB_CACHE_LINE_BYTES * 6)
```

Doesn't seem to happen on master (nor with the base branch - 22.02 so likely has to do with the ipfix addition.

Reply:
Sorry for the late reply.
Concerning `mobileIMSI` field: I had an impression that it must be stored using TBCD encoding, but will recheck
Concerning the segfault (actually SIGABRT I think due to heap exhaustion): you may try increasing the heap size, e.g. by setting `heapsize 3G` in startup.conf